### PR TITLE
Rewrite cloud network form additions

### PR DESCRIPTION
**Pull request by gubbe505**
Additional changes to the CloudNetwork dialog.

### CODE REVIEW

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx [hunk 1]
```diff
@@ -2,24 +2,30 @@ import React, { useState, useEffect } from 'react';
...
   const [{ isLoading, initialValues }, setState] = useState({
     isLoading: !!recordId,
-    initialValues: {},
+    initialValues: {},  
   });
 
   useEffect(() => {
-    if (recordId) {
-      API.get(`/api/cloud_networks/${recordId}`).then(({ id, providerId, ...initialValues }) => {
+    if (!!recordId && ManageIQ.controller === 'cloud_network') {
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx:13]: A CloudNetwork can be created from the Network -> CloudNetwork overview and and the Network > Provider > CloudNetwork overview. With the last one the recordId is also set by MIQ. Therefor we need to check the controller. @skateman If you know a better way of doing this, please do. I also think we should refactor this into a separate provider-form for reusing it for the other forms. Even better if this form is available in the ui repo.

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js [hunk 2]
```diff
@@ -26,45 +26,24 @@ const subnet_mask_options = [
     value: '/29',
   }
 ];
-const security_zone_options = [
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js:29]: Only for the KPN network.

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js [hunk 3]
```diff
@@ -26,45 +26,24 @@ const subnet_mask_options = [
...
 
-const loadProviders = () => API.get("/api/providers?collection_class=ManageIQ::Providers::Nsxt::NetworkManager").then(({ resources }) =>
+const loadProviders = () => API.get("/api/providers?collection_class=ManageIQ::Providers::Nsxt::NetworkManager&expand=resources&attributes=id,name").then(({ resources }) =>
   resources.map(({ id: value, name: label }) => ({ label, value }))
 );
 
 export default (edit) => {
   const schema = {
     fields: [
-      {...(edit ? {} : {
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js]: This construction give me an error.

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import { ProvidersApi } from '../../utils/providers.api';
+
+const NsxtCloudNetworkForm = ({ recordId }) => {
+  const [{ isLoading, initialValues }, setState] = useState({
+    isLoading: !!recordId,
+    initialValues: {},
+  });
+
+  useEffect(() => {
+    if (recordId) {
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx]: A CloudNetwork can be created from the Network -> CloudNetwork overview and and the Network > Provider > CloudNetwork overview. With the last one the recordId is also set by MIQ. 
Therefor we need to check the recordID and ManageIQ.controller to know if the form is in Edit mode. This also go for the submit function

 If you know a better way of doing this, please do. 
I also think we should refactor this into a separate provider-form for reusing it for the other forms. Even better if this form is available in the ui repo.

```
    if (!!recordId && ManageIQ.controller === 'cloud_network') {
      API.get(`/api/cloud_networks/${recordId}?attributes=name,description`).then(({ id, providerId, ...initialValues }) => {
        setState({
          isLoading: false,
          initialValues,
        });
      });
    }
    if (!!recordId &&  ManageIQ.controller === 'ems_network') {
      initialValues.ems_id = recordId;
      setState({
        isLoading: false,
        initialValues,
      });
```

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js [hunk 5]
```diff
@@ -49,9 +49,22 @@ const security_zone_options = [
   }
 ];
 
-export default () => {
+const loadProviders = () => API.get("/api/providers?collection_class=ManageIQ::Providers::Nsxt::NetworkManager").then(({ resources }) =>
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js]: We need some additional attributes...

```
const loadProviders = () => API.get("/api/providers?collection_class=ManageIQ::Providers::Nsxt::NetworkManager&expand=resources&attributes=id,name").then(({ resources }) =>
  resources.map(({ id: value, name: label }) => ({ label, value }))
);
````

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js [hunk 6]
```diff
@@ -49,9 +49,22 @@ const security_zone_options = [
...
+  resources.map(({ id: value, name: label }) => ({ label, value }))
+);
+
+export default (edit) => {
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.schema.js:34]: `...(edit ? {} : {` give me a error. Using `hideField: edit` works.

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    ManageIQ.redux.store.dispatch({
+      type: 'FormButtons.callbacks',
+      payload: { addClicked: () => formOptions.submit() },
+    });
+  };
+
+  const onSubmit = (values) => {
+    miqSparkleOn();
+    const request = recordId ? API.patch(`/api/cloud_networks/${recordId}`, values) : API.post('/api/cloud_networks', values);
```
Comment (gubbe505) [inline app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx]: We need to update the API to allow this patch/post API call.

This is working with the current API.
```
    const request = !!recordId && ManageIQ.controller === 'cloud_network' ? 
      API.post(`/api/providers/${values.ems_id}/cloud_networks`, {
        action: 'edit',
        id: values.id,
        resource: { ...values },
      }) : 
      API.post(`/api/providers/${values.ems_id}/cloud_networks`, {
        action: 'create',
        resource: values,
      });
```

File: app/javascript/packs/component-definitions-common.js [hunk 8]
```diff
@@ -1,6 +1,5 @@
-import CreateNsxtCloudNetworkForm from '../components/cloud-networks/create-nsxt-cloud-network-form';
+import NsxtCloudNetworkForm from '../components/cloud-networks/create-nsxt-cloud-network-form';
```
Comment (gubbe505) [inline app/javascript/packs/component-definitions-common.js]: import NsxtCloudNetworkForm from '../components/cloud-networks/nsxt-cloud-network-form';

**Review by gubbe505 — COMMENTED** [review]
In the `CloudNetworkCenter` and `CloudNetworkCenters` the component_name have to be changed to `NsxtCloudNetworkForm`.

File: app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx (hunk 1 above)
Comment (skateman) [reply app/javascript/components/cloud-networks/nsxt-cloud-network-form.jsx:13]: Are you saying we don't need to load the values for other forms when editing?

**Pull request closed without merging**